File: src/daytrace/markdown.py
from __future__ import annotations

import html
import re
from zoneinfo import ZoneInfo

from daytrace.diagnostics import diagnostic_messages
from daytrace.models import (
    ActivityRecord,
    ActivityReport,
    ActivitySession,
    ActivitySlice,
    ContextSignal,
    SessionBundle,
    SourceKind,
)


SOURCE_LABEL = {
    SourceKind.WINDOW: "Current window",
    SourceKind.EDITOR: "Editor",
    SourceKind.BROWSER: "Browser",
}


def format_duration(seconds: float) -> str:
    if 0 < seconds < 60:
        return "<1m"
    minutes = max(0, int((seconds + 30) // 60))
    hours, minutes = divmod(minutes, 60)
    if hours and minutes:
        return f"{hours}h {minutes}m"
    if hours:
        return f"{hours}h"
    return f"{minutes}m"


def _single_line(value: str) -> str:
    return " ".join(value.replace("\r", " ").replace("\n", " ").split())


def _escape(value: str) -> str:
    escaped = html.escape(_single_line(value), quote=False).replace("\\", "\\\\")
    return re.sub(r"([`*_{}\[\]#+|])", r"\\\1", escaped)


def _code(value: str) -> str:
    value = _single_line(value)
    longest = max((len(run) for run in re.findall(r"`+", value)), default=0)
    fence = "`" * (longest + 1)
    padding = " " if value.startswith("`") or value.endswith("`") else ""
    return f"{fence}{padding}{value}{padding}{fence}"


def _basename(value: str | None) -> str | None:
    return value.rsplit("/", 1)[-1].rsplit("\\", 1)[-1] if value else None


def _label(record: ActivityRecord) -> str:
    if record.kind is SourceKind.WINDOW:
        values = (record.app, record.title)
    elif record.kind is SourceKind.BROWSER:
        values = (record.url_host, record.title)
    else:
        values = (_basename(record.project), _basename(record.file), record.language)
    label = " — ".join(_escape(value) for value in values if value)
    return label or SOURCE_LABEL[record.kind]


def render_markdown(report: ActivityReport) -> str:
    lines = [f"# Activity summary — {report.day.isoformat()}", ""]
    if report.project is not None:
        lines.append(f"Project filter: {_code(report.project)}")
    lines.extend([f"Timezone: {_code(report.timezone_name)}", "", "## Overview", ""])
    if not report.timeline:
        return "\n".join(lines + ["No matching activity.", ""])

    active = (
        format_duration(report.active_seconds)
        if report.active_seconds is not None
        else "Unavailable"
    )
    lines.extend(
        [
            f"- Active matched time: {active}",
            f"- Timeline entries: {len(report.timeline)}",
            f"- Sources: {', '.join(SOURCE_LABEL[item] for item in report.sources)}",
            "",
            "## Timeline",
            "",
        ]
    )
    zone = ZoneInfo(report.timezone_name)
    for record in report.timeline:
        start = record.start.astimezone(zone).strftime("%H:%M")
        end = record.end.astimezone(zone).strftime("%H:%M")
        lines.append(
            f"- {start}–{end} ({format_duration(record.duration_seconds)}) — {_label(record)}"
        )
    if report.applications:
        lines.extend(["", "## Applications", ""])
        lines.extend(
            f"- {_escape(item.app)} — {format_duration(item.seconds)}"
            for item in report.applications
        )
    return "\n".join(lines) + "\n"


def _session_times(item: ActivitySession | ActivitySlice, zone: ZoneInfo) -> str:
    start = item.start.astimezone(zone).strftime("%H:%M")
    end = item.end.astimezone(zone).strftime("%H:%M")
    return f"{start}–{end}"


def _context_label(context: ContextSignal) -> str:
    values = (
        context.project,
        context.file,
        context.url_host,
        context.url_path,
        context.title,
        context.language,
    )
    return " — ".join(_escape(value) for value in values if value) or context.kind.value


def _append_session_details(
    lines: list[str], item: ActivitySession, zone: ZoneInfo
) -> None:
    if item.evidence_ids:
        lines.append(f"  - Evidence: {', '.join(_code(value) for value in item.evidence_ids)}")
    contexts = {
        (context.kind.value, _context_label(context))
        for activity_slice in item.slices
        for context in activity_slice.contexts
    }
    for kind, label in sorted(contexts):
        lines.append(f"  - {SOURCE_LABEL[SourceKind(kind)]}: {label}")
    if item.focused_seconds is None:
        lines.append("  - Foreground duration: unavailable")
    else:
        lines.append(f"  - Foreground duration: {format_duration(item.focused_seconds)}")


def _append_raw_slices(
    lines: list[str], item: ActivitySession, zone: ZoneInfo
) -> None:
    for activity_slice in item.slices:
        label = activity_slice.title or activity_slice.app or "Context evidence"
        focus = "focused" if activity_slice.focused else "evidence only"
        lines.append(
            f"  - {_session_times(activity_slice, zone)} — {_code(label)} "
            f"({format_duration(activity_slice.duration_seconds)}, {focus})"
        )


def render_session_markdown(
    bundle: SessionBundle, *, details: bool = False, raw: bool = False
) -> str:
    zone = ZoneInfo(bundle.timezone_name)
    focused = (
        format_duration(bundle.focused_seconds)
        if bundle.focused_seconds is not None
        else "Unavailable"
    )
    lines = [
        f"# DayTrace — {bundle.day.isoformat()}",
        "",
        f"Timezone: {_code(bundle.timezone_name)} (inferred at query time)",
        f"Focused activity: {focused}",
        "Summary: Deterministic activity sessions",
        "",
        "## Activity sessions",
        "",
    ]
    if not bundle.sessions:
        lines.append("No activity sessions.")
    for item in sorted(bundle.sessions, key=lambda value: (value.start, value.session_id)):
        lines.append(
            f"- {_session_times(item, zone)} — {_code(item.label)} "
            f"({format_duration(item.active_seconds)})"
        )
        for signal in item.outcome_signals:
            lines.append(f"  - Observed signal: {_escape(signal.label)}")
        if details:
            _append_session_details(lines, item, zone)
        if raw:
            _append_raw_slices(lines, item, zone)

    if bundle.diagnostics:
        lines.extend(["", "## Diagnostics", ""])
        for message in diagnostic_messages(bundle.diagnostics):
            lines.append(f"- {message[0].upper() + message[1:]}.")
    return "\n".join(lines) + "\n"

File: packages/daytrace-core/src/markdown.ts
import { diagnosticMessages } from "./diagnostics.js";
import type {
  ActivityEpisode,
  ActivityRecord,
  ActivityReport,
  ActivitySession,
  ActivitySlice,
  ContextSignal,
  EpisodeBundle,
  SessionBundle,
  SummaryProvenance,
  SourceKind,
  WorkstreamDigest,
  WorkstreamSummary,
} from "./models.js";
import { durationSeconds, timestampMs } from "./time.js";

const SOURCE_LABEL: Readonly<Record<SourceKind, string>> = {
  "current-window": "Current window",
  editor: "Editor",
  browser: "Browser",
  afk: "AFK",
};

export function formatDuration(seconds: number): string {
  if (seconds > 0 && seconds < 60) return "<1m";
  const totalMinutes = Math.max(0, Math.floor((seconds + 30) / 60));
  const hours = Math.floor(totalMinutes / 60);
  const minutes = totalMinutes % 60;
  if (hours > 0 && minutes > 0) return `${hours}h ${minutes}m`;
  if (hours > 0) return `${hours}h`;
  return `${minutes}m`;
}

function singleLine(value: string): string {
  return value.replace(/[\r\n]/g, " ").trim().split(/\s+/u).filter(Boolean).join(" ");
}

function htmlEscape(value: string): string {
  return value.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
}

export function escapeMarkdown(value: string): string {
  return htmlEscape(singleLine(value)).replace(/\\/g, "\\\\").replace(/([`*_{}\[\]#+|])/g, "\\$1");
}

export function inlineCode(value: string): string {
  const clean = singleLine(value);
  const runs = clean.match(/`+/g) ?? [];
  const longest = runs.reduce((length, run) => Math.max(length, run.length), 0);
  const fence = "`".repeat(longest + 1);
  const padding = clean.startsWith("`") || clean.endsWith("`") ? " " : "";
  return `${fence}${padding}${clean}${padding}${fence}`;
}

function basename(value?: string): string | undefined {
  return value?.split("/").at(-1)?.split("\\").at(-1);
}

function recordLabel(record: ActivityRecord): string {
  const values = record.kind === "current-window"
    ? [record.app, record.title]
    : record.kind === "browser"
      ? [record.urlHost, record.title]
      : [basename(record.project), basename(record.file), record.language];
  return values.filter((value): value is string => value !== undefined).map(escapeMarkdown).join(" — ")
    || SOURCE_LABEL[record.kind];
}

function localTime(value: string, timezoneName: string): string {
  return new Intl.DateTimeFormat("en-GB-u-hc-h23", {
    timeZone: timezoneName,
    hour: "2-digit",
    minute: "2-digit",
    hourCycle: "h23",
  }).format(new Date(timestampMs(value)));
}

function timeRange(item: { readonly start: string; readonly end: string }, timezoneName: string): string {
  return `${localTime(item.start, timezoneName)}–${localTime(item.end, timezoneName)}`;
}

export function renderMarkdown(report: ActivityReport): string {
  const lines = [`# Activity summary — ${report.day}`, ""];
  if (report.project !== undefined) lines.push(`Project filter: ${inlineCode(report.project)}`);
  lines.push(`Timezone: ${inlineCode(report.timezoneName)}`, "", "## Overview", "");
  if (report.timeline.length === 0) return `${[...lines, "No matching activity.", ""].join("\n")}`;
  lines.push(
    `- Active matched time: ${report.activeSeconds === undefined ? "Unavailable" : formatDuration(report.activeSeconds)}`,
    `- Timeline entries: ${report.timeline.length}`,
    `- Sources: ${report.sources.map((item) => SOURCE_LABEL[item]).join(", ")}`,
    "",
    "## Timeline",
    "",
  );
  for (const record of report.timeline) {
    lines.push(`- ${timeRange(record, report.timezoneName)} (${formatDuration(durationSeconds(record.start, record.end))}) — ${recordLabel(record)}`);
  }
  if (report.applications.length > 0) {
    lines.push("", "## Applications", "");
    for (const item of report.applications) lines.push(`- ${escapeMarkdown(item.app)} — ${formatDuration(item.seconds)}`);
  }
  return `${lines.join("\n")}\n`;
}

function contextLabel(context: ContextSignal): string {
  const values = [context.project, context.file, context.urlHost, context.urlPath, context.title, context.language];
  return values.filter((value): value is string => value !== undefined).map(escapeMarkdown).join(" — ") || context.kind;
}

function appendSessionDetails(lines: string[], item: ActivitySession): void {
  if (item.evidenceIds.length > 0) lines.push(`  - Evidence: ${item.evidenceIds.map(inlineCode).join(", ")}`);
  const contexts = new Map<string, readonly [SourceKind, string]>();
  for (const context of item.slices.flatMap((slice) => slice.contexts)) {
    const value = contextLabel(context);
    contexts.set(`${context.kind}\0${value}`, [context.kind, value]);
  }
  for (const [, [kind, value]] of [...contexts].sort(([left], [right]) => left.localeCompare(right))) {
    lines.push(`  - ${SOURCE_LABEL[kind]}: ${value}`);
  }
  lines.push(`  - Foreground duration: ${item.focusedSeconds === undefined ? "unavailable" : formatDuration(item.focusedSeconds)}`);
}

function appendRawSlices(lines: string[], item: ActivitySession, timezoneName: string): void {
  for (const slice of item.slices) {
    const label = slice.title ?? slice.app ?? "Context evidence";
    const focus = slice.focused ? "focused" : "evidence only";
    lines.push(`  - ${timeRange(slice, timezoneName)} — ${inlineCode(label)} (${formatDuration(durationSeconds(slice.start, slice.end))}, ${focus})`);
  }
}

export interface MarkdownDetailsOptions {
  readonly details?: boolean;
  readonly raw?: boolean;
}

export function renderSessionMarkdown(bundle: SessionBundle, options: MarkdownDetailsOptions = {}): string {
  const focused = bundle.focusedSeconds === undefined ? "Unavailable" : formatDuration(bundle.focusedSeconds);
  const lines = [
    `# DayTrace — ${bundle.day}`,
    "",
    `Timezone: ${inlineCode(bundle.timezoneName)} (inferred at query time)`,
    `Focused activity: ${focused}`,
    "Summary: Deterministic activity sessions",
    "",
    "## Activity sessions",
    "",
  ];
  const sessions = [...bundle.sessions].sort((left, right) => timestampMs(left.start) - timestampMs(right.start) || left.sessionId.localeCompare(right.sessionId));
  if (sessions.length === 0) lines.push("No activity sessions.");
  for (const item of sessions) {
    lines.push(`- ${timeRange(item, bundle.timezoneName)} — ${inlineCode(item.label)} (${formatDuration(item.activeSeconds)})`);
    for (const signal of item.outcomeSignals) lines.push(`  - Observed signal: ${escapeMarkdown(signal.label)}`);
    if (options.details) appendSessionDetails(lines, item);
    if (options.raw) appendRawSlices(lines, item, bundle.timezoneName);
  }
  if (bundle.diagnostics.length > 0) {
    lines.push("", "## Diagnostics", "");
    for (const message of diagnosticMessages(bundle.diagnostics)) lines.push(`- ${message[0]?.toLocaleUpperCase()}${message.slice(1)}.`);
  }
  return `${lines.join("\n")}\n`;
}

function episodeLine(item: ActivityEpisode, timezoneName: string): string {
  return `- ${timeRange(item, timezoneName)} — ${inlineCode(item.label)} (${formatDuration(item.activeSeconds)})`;
}

export function renderEpisodeMarkdown(bundle: EpisodeBundle, options: MarkdownDetailsOptions = {}): string {
  const focused = bundle.focusedSeconds === undefined ? "Unavailable" : formatDuration(bundle.focusedSeconds);
  const lines = [
    `# DayTrace — ${bundle.day}`,
    "",
    `Timezone: ${inlineCode(bundle.timezoneName)} (inferred at query time)`,
    `Focused activity: ${focused}`,
    "Summary: Deterministic activity episodes",
    "",
    "## Activity episodes",
    "",
  ];
  const episodes = [...bundle.episodes].sort((left, right) => timestampMs(left.start) - timestampMs(right.start) || left.episodeId.localeCompare(right.episodeId));
  if (episodes.length === 0) lines.push("No activity episodes.");
  for (const item of episodes) {
    lines.push(episodeLine(item, bundle.timezoneName));
    if (item.anchors.length > 0) lines.push(`  - Anchors: ${item.anchors.map((anchor) => `${anchor.kind}: ${escapeMarkdown(anchor.value)}`).join("; ")}`);
    if (item.applications.length > 0) lines.push(`  - Tools: ${item.applications.map((value) => `${escapeMarkdown(value.value)} ×${value.count}`).join(", ")}`);
    lines.push(`  - Activity transitions: ${item.sessionIds.length}`);
    for (const signal of item.outcomeSignals) lines.push(`  - Observed signal: ${escapeMarkdown(signal.label)}`);
    if (options.details) {
      lines.push(`  - Episode ID: ${inlineCode(item.episodeId)}`);
      lines.push(`  - Source sessions: ${item.sessionIds.map(inlineCode).join(", ")}`);
      if (item.activityLabels.length > 0) lines.push(`  - Activity labels: ${item.activityLabels.map((value) => `${escapeMarkdown(value.value)} ×${value.count}`).join(", ")}`);
      if (item.evidenceIds.length > 0) lines.push(`  - Evidence: ${item.evidenceIds.map(inlineCode).join(", ")}`);
      lines.push(`  - Foreground duration: ${item.focusedSeconds === undefined ? "unavailable" : formatDuration(item.focusedSeconds)}`);
    }
  }
  if (options.raw) {
    lines.push("", "## Fine-grained activity", "");
    for (const item of [...bundle.sessions].sort((left, right) => timestampMs(left.start) - timestampMs(right.start) || left.sessionId.localeCompare(right.sessionId))) {
      lines.push(`- ${timeRange(item, bundle.timezoneName)} — ${inlineCode(item.label)} (${formatDuration(item.activeSeconds)}) — ${inlineCode(item.sessionId)}`);
      appendSessionDetails(lines, item);
      appendRawSlices(lines, item, bundle.timezoneName);
    }
  }
  if (bundle.diagnostics.length > 0) {
    lines.push("", "## Diagnostics", "");
    for (const message of diagnosticMessages(bundle.diagnostics)) lines.push(`- ${message[0]?.toLocaleUpperCase()}${message.slice(1)}.`);
  }
  return `${lines.join("\n")}\n`;
}

function outcomeCell(item: WorkstreamSummary): string {
  const visible = item.outcomes.filter((outcome) => outcome.strength !== "none");
  if (visible.length === 0) return "No completion should be claimed from this trace alone.";
  return visible.map((outcome) => `${outcome.strength === "likely" ? "Likely: " : ""}${escapeMarkdown(outcome.text)}`).join("<br>");
}

function topicCell(item: WorkstreamSummary): string {
  return item.topics.length === 0
    ? "No specific topics identified."
    : item.topics.map((topic) => escapeMarkdown(topic.text)).join("<br>");
}

export interface DigestMarkdownOptions {
  readonly details?: boolean;
}

export function renderDigestMarkdown(
  bundle: EpisodeBundle,
  digest: WorkstreamDigest,
  provenance: SummaryProvenance,
  options: DigestMarkdownOptions = {},
): string {
  const focused = bundle.focusedSeconds === undefined ? "Unavailable" : formatDuration(bundle.focusedSeconds);
  const episodeById = new Map(bundle.episodes.map((item) => [item.episodeId, item]));
  const lines = [
    `# DayTrace — ${bundle.day}`,
    "",
    `Timezone: ${inlineCode(bundle.timezoneName)} (inferred at query time)`,
    `Focused activity: ${focused}`,
    `Summary: AI-assisted workstreams (${inlineCode(provenance.provider)} / ${inlineCode(provenance.model)})`,
  ];
  if (options.details) lines.push(`Provider requests: ${provenance.requestCount}`);
  lines.push(
    "",
    "## What this day appears to contain",
    "",
    "This is a confidence-aware summary inferred from the activity trace. Achievements require evidence of a resulting state, not merely an open application.",
    "",
    "| Project / workstream | Apparent achievements | Work and topics |",
    "| --- | --- | --- |",
  );
  for (const workstream of digest.workstreams) {
    lines.push(`| **${escapeMarkdown(workstream.label)}** | ${outcomeCell(workstream)} | ${topicCell(workstream)} |`);
  }
  if (digest.workstreams.length === 0) lines.push("| _No coherent workstreams identified._ | — | — |");
  if (!options.details) return `${lines.join("\n")}\n`;
  lines.push("", "## Supporting activity", "");
  const confidence = { high: "High", medium: "Medium", low: "Low" } as const;
  for (const workstream of digest.workstreams) {
    const episodes = workstream.episodeIds.map((id) => episodeById.get(id)!);
    lines.push(
      `### ${escapeMarkdown(workstream.label)}`,
      "",
      `Primary allocation · ${confidence[workstream.confidence]} confidence · ${formatDuration(episodes.reduce((total, item) => total + item.activeSeconds, 0))}`,
      "",
    );
    for (const episode of [...episodes].sort((left, right) => timestampMs(left.start) - timestampMs(right.start) || left.episodeId.localeCompare(right.episodeId))) {
      lines.push(episodeLine(episode, bundle.timezoneName));
      lines.push(`  - Episode ID: ${inlineCode(episode.episodeId)}`);
      lines.push(`  - Activity transitions: ${episode.sessionIds.length}`);
    }
  }
  lines.push("", "## Unassigned activity", "");
  if (digest.unassignedEpisodeIds.length === 0) lines.push("None.");
  else for (const id of digest.unassignedEpisodeIds) {
    const episode = episodeById.get(id)!;
    lines.push(episodeLine(episode, bundle.timezoneName));
    lines.push(`  - Episode ID: ${inlineCode(episode.episodeId)}`);
  }
  return `${lines.join("\n")}\n`;
}
